### PR TITLE
[Bundle products in order] Move validation error to a notice at the bottom

### PR DESCRIPTION
**Pull request by jaclync**
Part of #11259
⚠️ Please review https://github.com/woocommerce/woocommerce-ios/pull/11292 first

## Why

From the design review w3JBReW1UJesCf02sj43kR-fi-984_9201, we'd like to move any validation error from either the top (bundle level) or inline (bundle item level) error message to a notice at the bottom.

The design looks like:
Screenshot 2023-11-28 at 4 05 14 PM

Please note that there are still some further enhancements we can make to match the design 100%:
- Update the bundle-level error message to match the design like `This bundle requires a total count of 2. Add 1 more of any product to continue.`. I personally found it a bit verbose and would like to get your thought on whether it's worth making this change
- Hide the notice when the bundle becomes valid, like after 1 second - I'll think about how to implement this

## How

A notice view is shown when:

- The bundle configuration is invalid
- The bundle configuration **becomes** valid
  - In order to implement this state change, I added `Publisher.withPrevious` publisher to observe the current & previous values of the validation state

In `ConfigurableBundleItemView` and `ConfigurableBundleProductView`, the error message text was removed. The new notice view is `ConfigurableBundleNoticeView`, which depends on a validation state `Result<Void, ValidationError>` from the view model `ConfigurableBundleProductViewModel`. I chose to only show the first error message so that the notice height doesn't take up a lot of vertical space when there are multiple errors.

For each item level validation error message in `ConfigurableBundleItemViewModel`, the product name is now included for the merchant to know which item to update.

## Testing instructions

Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product **with a bundle size rule** and **at least 1 optional variable item** with non-empty variations and **at least 1 non-variable item**.

- Go to the Orders tab
- Tap `+` to create an order
- Tap `Add Products` and select the bundle product in the prerequisite --> a notice should be shown if the default quantities don't match the bundle size rule
- Configure the non-variable items so that they are valid --> the Save CTA should be enabled, and a green notice should be shown
- Tap the circle to select the variable item --> the Save CTA should be disabled while a notice with an error message appears
- Tap `Choose variation` to choose a variation
- In the configuration form, the attributes should be shown that look like the expected design
- Configure the variable item so that it's valid --> the Save CTA should be enabled, and a green notice should be shown
- Tap `Save Configuration`
- Tap `1 Product Selected` --> the bundle product should be added to the order with the expected configuration

## Screenshots

bundle level error | valid state - light mode | item level error | valid state - dark mode
-- | -- | -- | --
 |  |  | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift [hunk 1]
```diff
@@ -268,35 +268,39 @@ private extension ConfigurableBundleItemViewModel {
...
         enum ErrorMessage {
-            static let invalidQuantityWithMinQuantityRule = NSLocalizedString(
-                "configureBundleItemError.invalidQuantityWithMinQuantityRule",
-                value: "The quantity needs to be at least %1$@",
-                comment: "Error message when setting a bundle item's quantity with a minimum quantity rule."
+            static let invalidQuantityWithMinQuantityRuleFormat = NSLocalizedString(
+                "configureBundleItemError.invalidQuantityWithMinQuantityRuleFormat",
+                value: "The quantity of %1$@ needs to be at least %2$@.",
+                comment: "Error message when setting a bundle item's quantity with a minimum quantity rule. " +
+                "%1$@ is the variable product name. %2$@ is the minimum quantity."
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift]: super nit: This applies to all products, not just variable products:

```suggestion
                "%1$@ is the product name. %2$@ is the minimum quantity."
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift [hunk 2]
```diff
@@ -268,35 +268,39 @@ private extension ConfigurableBundleItemViewModel {
...
             )
-            static let invalidQuantityWithMinAndMaxQuantityRules = NSLocalizedString(
-                "configureBundleItemError.invalidQuantityWithMinAndMaxQuantityRules",
-                value: "The quantity needs to be within %1$@ and %2$@",
-                comment: "Error message when setting a bundle item's quantity with minimum and maximum quantity rules."
+            static let invalidQuantityWithMinAndMaxQuantityRulesFormat = NSLocalizedString(
+                "configureBundleItemError.invalidQuantityWithMinAndMaxQuantityRulesFormat",
+                value: "The quantity of %1$@ needs to be within %2$@ and %3$@.",
+                comment: "Error message when setting a bundle item's quantity with minimum and maximum quantity rules. " +
+                "%1$@ is the variable product name. %2$@ is the minimum quantity, and %3$@ is the maximum quantity"
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift]: super nit: Same as above:

```suggestion
                "%1$@ is the product name. %2$@ is the minimum quantity, and %3$@ is the maximum quantity"
```

File: WooCommerce/Classes/Extensions/Publisher+WithPrevious.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    /// Includes the current element as well as the previous element from the upstream publisher in a tuple where the previous element is optional.
+    /// The first time the upstream publisher emits an element, the previous element will be `nil`.
+    /// Reference: https://stackoverflow.com/a/67133582/9185596
+    ///
+    /// - Returns: A publisher of a tuple of the previous and current elements from the upstream publisher.
+    func withPrevious() -> AnyPublisher<(previous: Output?, current: Output), Failure> {
+        scan(Optional<(Output?, Output)>.none) { ($0?.1, $1) }
+            .compactMap { $0 }
+            .eraseToAnyPublisher()
+    }
```
Comment (rachelmcr) [inline WooCommerce/Classes/Extensions/Publisher+WithPrevious.swift:13]: Nice! ⭐

**Review by rachelmcr — APPROVED** [review]
These changes look good! The styling and placement of this notice are really nice. 🎉 

> Update the bundle-level error message to match the design like `This bundle requires a total count of 2. Add 1 more of any product to continue.`. I personally found it a bit verbose and would like to get your thought on whether it's worth making this change

I agree, that's fairly verbose (and is likely to be even longer in some languages). The current message (`Please choose 2-3 items.` or `Please choose at least 2 items.`) seems clear but concise. It also fits with the customer-facing messaging about bundle quantity used on the front of the store.

> Hide the notice when the bundle becomes valid, like after 1 second

That would be really nice. It does seem like that valid configuration message is a little redundant, now that the "Save Configuration" button is also at the bottom and only becomes active when the configuration is valid.

_Unrelated validation issue:_

Also, this is not specific to these changes but I happened to choose a variation that triggered a validation issue. If I choose a variation that has an "any" attribute option, I can't get the configuration to become valid. No matter what, it persistently tells me that I need to choose an option for all variation attributes for that product:

https://github.com/woocommerce/woocommerce-ios/assets/8658164/c415a237-5b98-4b57-9ab7-a2faefcea467

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift]: Updated in https://github.com/woocommerce/woocommerce-ios/pull/11306/commits/d0d6dcfbdd2e8d94bc0283004a37df8540d13d00

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ConfigurableBundleItemViewModel.swift]: Good catch, updated in https://github.com/woocommerce/woocommerce-ios/pull/11306/commits/d0d6dcfbdd2e8d94bc0283004a37df8540d13d00

**Comment by jaclync** [comment]
> I agree, that's fairly verbose (and is likely to be even longer in some languages). The current message (`Please choose 2-3 items.` or `Please choose at least 2 items.`) seems clear but concise. It also fits with the customer-facing messaging about bundle quantity used on the front of the store.

Sounds good, I'll keep the message as it is.

> That would be really nice. It does seem like that valid configuration message is a little redundant, now that the "Save Configuration" button is also at the bottom and only becomes active when the configuration is valid.

I'll see if I can get to this after the time zone task https://github.com/woocommerce/woocommerce-ios/issues/10842.

> _Unrelated validation issue:_
> 
> Also, this is not specific to these changes but I happened to choose a variation that triggered a validation issue. If I choose a variation that has an "any" attribute option, I can't get the configuration to become valid. No matter what, it persistently tells me that I need to choose an option for all variation attributes for that product:

Thanks for reporting this! I was able to repro this on my own test site. I'll look into this and make a fix separately.

**Pull request merged**